### PR TITLE
Data binding and more

### PR DESCRIPTION
**Pull request by willmcgugan**
Bit of a mixed bag. Added a number of things while working on Tailless. The most notable is probably the data binding.

See changelog for details.

### CODE REVIEW

File: docs/guide/reactivity.md [hunk 1]
```diff
@@ -266,3 +266,140 @@ When the result of `compute_color` changes, Textual will also call `watch_color`
...
+
+    ```{.textual path="docs/examples/guide/reactivity/world_clock02.py"}
+    ```
+
+Note how the addition of the `data_bind` methods negates the need for the watcher in `world_clock01.py`.
+
+
+!!! note
+
+    Data bindings works in a single direction.
```
Comment (darrenburns) [inline docs/guide/reactivity.md:376]:
```suggestion
    Data binding works in a single direction.
```

**Review by darrenburns — APPROVED** [review]
LGTM

File: src/textual/dom.py [hunk 2]
```diff
@@ -194,9 +197,127 @@ def __init__(
...
+        See the [guide on data binding](/guide/reactivity#data-binding).
+
+        Example:
+            ```python
+            def compose(self) -> ComposeResult:
+                yield WorldClock("Europe/London").data_bind(WorldClockApp.time)
+                yield WorldClock("Europe/Paris").data_bind(WorldClockApp.time)
+                yield WorldClock("Asia/Tokyo").data_bind(WorldClockApp.time)
+            ```
+
```
Comment (rodrigogiraoserrao) [inline src/textual/dom.py:250]:
```suggestion
```

File: src/textual/dom.py [hunk 3]
```diff
@@ -194,9 +197,127 @@ def __init__(
...
+                )
+            self._reactive_connect[name] = (parent, reactive)
+        self._initialize_data_bind()
+        return self
+
+    def _initialize_data_bind(self) -> None:
+        """initialize a data binding.
+
+        Args:
+            compose_parent: The node doing the binding.
```
Comment (rodrigogiraoserrao) [inline src/textual/dom.py:284]:
```suggestion
```

File: src/textual/dom.py [hunk 4]
```diff
@@ -194,9 +197,127 @@ def __init__(
...
+                    A callable which takes the value to set.
+                """
+
+                def setter(value: object) -> None:
+                    """Set bound data."""
+                    _rich_traceback_omit = True
+                    Reactive._initialize_object(self)
+                    setattr(self, variable_name, value)
+
+                return setter
```
Comment (rodrigogiraoserrao) [inline src/textual/dom.py:306]: Can't this be defined outside of the `for` loop?

File: src/textual/dom.py [hunk 5]
```diff
@@ -194,9 +197,127 @@ def __init__(
...
+
+            assert isinstance(compose_parent, DOMNode)
+            setter = make_setter(variable_name)
+            if isinstance(reactive, Reactive):
+                self.watch(
+                    compose_parent,
+                    reactive.name,
+                    setter,
```
Comment (rodrigogiraoserrao) [inline src/textual/dom.py:314]: Why not `partial(setattr, self, variable_name)`?

**Review by rodrigogiraoserrao — COMMENTED** [review]
I left a couple of questions for possible improvements.

File: src/textual/css/query.py [hunk 6]
```diff
@@ -444,3 +448,32 @@ def blur(self) -> DOMQuery[QueryType]:
...
                 self._node.screen._reset_focus(focused, avoiding=nodes)
         return self
+
+    def set(
+        self,
+        display: bool | None = None,
+        visible: bool | None = None,
+        disabled: bool | None = None,
+        loading: bool | None = None,
+    ) -> DOMQuery[QueryType]:
```
[the remark this answers is not in the document]
Comment (willmcgugan) [reply src/textual/css/query.py:458]: It's for the benefit of typechecking mainly.

**Pull request merged**